src/hotspot/share/opto/mempointer.cpp

```diff
@@ -0,0 +1,299 @@
+      // be truncated back with ConvL2I. On 64bit systems this is not linear:
+      //
+      //   ConvI2L(ConvL2I(max_jint + 1)) = ConvI2L(min_jint) = min_jint
+      //
+      NOT_LP64( case Op_ConvL2I: )
+      {
+        // Decompose: look through.
+        Node* a = n->in(1);
+        _worklist.push(MemPointerSummand(a, scale LP64_ONLY( COMMA scaleL )));
+        return;
+      }
+    }
+  }
+
+  // Default: we could not parse the "summand" further, i.e. it is terminal.
+  _summands.push(summand);
+}
+
+bool MemPointerLinearFormParser::is_safe_from_int_overflow(const int opc LP64_ONLY( COMMA const NoOverflowInt scaleL )) const {
+#ifndef _LP64
+  // On 32-bit platforms, ... TODO
+  return true;
+#else
+
+  // Not trivially safe:
+  //   AddI:     ConvI2L(a +  b)    != ConvI2L(a) +  ConvI2L(b)
+  //   SubI:     ConvI2L(a -  b)    != ConvI2L(a) -  ConvI2L(b)
+  //   MulI:     ConvI2L(a *  conI) != ConvI2L(a) *  ConvI2L(conI)
+  //   LShiftI:  ConvI2L(a << conI) != ConvI2L(a) << ConvI2L(conI)
+  //
+  // But these are always safe:
+  switch(opc) {
+    case Op_ConI:
+    case Op_ConL:
+    case Op_AddP:
+    case Op_AddL:
+    case Op_SubL:
+    case Op_MulL:
+    case Op_LShiftL:
+    case Op_CastII:
+    case Op_CastLL:
+    case Op_CastX2P:
+    // TODO CastPP ?
+    case Op_ConvI2L:
+      return true;
+  }
+
+  // TODO tests with native memory, etc.
+
+  const TypeAryPtr* ary_ptr_t = _mem->adr_type()->isa_aryptr();
+  if (ary_ptr_t != nullptr) {
+    // Array accesses that are not Unsafe always have a RangeCheck which ensures
+    // that there is no int overflow.
+    if (!_mem->is_unsafe_access()) {
+      return true;
+    }
+
+    // TODO
+    BasicType array_element_bt = ary_ptr_t->elem()->array_element_basic_type();
+    if (is_java_primitive(array_element_bt)) {
+      NoOverflowInt array_element_size_in_bytes = NoOverflowInt(type2aelembytes(array_element_bt));
+      if (scaleL.is_multiple_of(array_element_size_in_bytes)) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+#endif
+}
+
+MemPointerAliasing MemPointerLinearForm::get_aliasing_with(const MemPointerLinearForm& other
+                                                           NOT_PRODUCT( COMMA const TraceMemPointer& trace) ) const {
+#ifndef PRODUCT
+  if (trace.is_trace_aliasing()) {
+    tty->print_cr("MemPointerLinearForm::get_aliasing_with:");
+    print_on(tty);
+    other.print_on(tty);
+  }
+#endif
+
+  // Check if all summands are the same:
+  for (uint i = 0; i < SUMMANDS_SIZE; i++) {
+    const MemPointerSummand s1 = summands_at(i);
+    const MemPointerSummand s2 = other.summands_at(i);
+    if (s1 != s2) {
+#ifndef PRODUCT
+      if (trace.is_trace_aliasing()) {
+        tty->print_cr("  -> Aliasing unknown, differ on summand %d.", i);
+      }
+#endif
+      return MemPointerAliasing::make_unknown();
+    }
+  }
+
+  // Compute distance:
+  NoOverflowInt distance = other.con() - con();
+  distance = distance.truncate_to_30_bits();
```

Comment:
naming could be an issue, why coud it be NaN

Reply:
Ok, I now replaced it with a `bool` method `is_abs_less_than_2_to_30`. I think that is more clear.